File: src/main/java/de/storchp/opentracks/osmplugin/dashboardapi/Waypoint.java
package de.storchp.opentracks.osmplugin.dashboardapi;

import static de.storchp.opentracks.osmplugin.dashboardapi.APIConstants.LAT_LON_FACTOR;

import android.content.ContentResolver;
import android.database.Cursor;
import android.net.Uri;

import org.mapsforge.core.model.LatLong;

import java.util.ArrayList;
import java.util.List;

import de.storchp.opentracks.osmplugin.utils.MapUtils;

public class Waypoint {

    public static final String _ID = "_id";
    public static final String NAME = "name"; // waypoint name
    public static final String DESCRIPTION = "description"; // waypoint description
    public static final String CATEGORY = "category"; // waypoint category
    public static final String ICON = "icon"; // waypoint icon
    public static final String TRACKID = "trackid"; // track id
    public static final String LONGITUDE = "longitude"; // longitude
    public static final String LATITUDE = "latitude"; // latitude
    public static final String PHOTOURL = "photoUrl"; // url for the photo

    public static final String[] PROJECTION = {
            _ID,
            NAME,
            DESCRIPTION,
            CATEGORY,
            ICON,
            TRACKID,
            LATITUDE,
            LONGITUDE,
            PHOTOURL
    };

    private final long id;
    private final String name;
    private final String description;
    private final String category;
    private final String icon;
    private final long trackId;
    private final LatLong latLong;
    private final String photoUrl;

    public Waypoint(long id, String name, String description, String category, String icon, long trackId, LatLong latLong, String photoUrl) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.category = category;
        this.icon = icon;
        this.trackId = trackId;
        this.latLong = latLong;
        this.photoUrl = photoUrl;
    }

    /**
     * Reads the Waypoints from the Content Uri.
     */
    public static List<Waypoint> readWaypoints(ContentResolver resolver, Uri data, long lastWaypointId) {
        var waypoints = new ArrayList<Waypoint>();
        try (Cursor cursor = resolver.query(data, Waypoint.PROJECTION, null, null, null)) {
            while (cursor.moveToNext()) {
                var waypointId = cursor.getLong(cursor.getColumnIndexOrThrow(Waypoint._ID));
                if (lastWaypointId > 0 && lastWaypointId >= waypointId) { // skip waypoints we already have
                    continue;
                }
                var name = cursor.getString(cursor.getColumnIndexOrThrow(Waypoint.NAME));
                var description = cursor.getString(cursor.getColumnIndexOrThrow(Waypoint.DESCRIPTION));
                var category = cursor.getString(cursor.getColumnIndexOrThrow(Waypoint.CATEGORY));
                var icon = cursor.getString(cursor.getColumnIndexOrThrow(Waypoint.ICON));
                var trackId = cursor.getLong(cursor.getColumnIndexOrThrow(Waypoint.TRACKID));
                var latitude = cursor.getInt(cursor.getColumnIndexOrThrow(Waypoint.LATITUDE)) / LAT_LON_FACTOR;
                var longitude = cursor.getInt(cursor.getColumnIndexOrThrow(Waypoint.LONGITUDE)) / LAT_LON_FACTOR;
                if (MapUtils.isValid(latitude, longitude)) {
                    var latLong = new LatLong(latitude, longitude);
                    var photoUrl = cursor.getString(cursor.getColumnIndexOrThrow(Waypoint.PHOTOURL));
                    waypoints.add(new Waypoint(waypointId, name, description, category, icon, trackId, latLong, photoUrl));
                }
            }
        }
        return waypoints;
    }


    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public String getCategory() {
        return category;
    }

    public String getIcon() {
        return icon;
    }

    public long getTrackId() {
        return trackId;
    }

    public LatLong getLatLong() {
        return latLong;
    }

    public String getPhotoUrl() {
        return photoUrl;
    }
}

File: src/main/kotlin/de/storchp/opentracks/osmplugin/dashboardapi/Waypoint.kt
package de.storchp.opentracks.osmplugin.dashboardapi

import android.content.ContentResolver
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import de.storchp.opentracks.osmplugin.map.MapData
import de.storchp.opentracks.osmplugin.map.MapUtils
import org.oscim.core.GeoPoint
import java.net.URLDecoder
import java.util.regex.Pattern

data class Waypoint(
    val id: Long = 0,
    val name: String?,
    val description: String? = null,
    val category: String? = null,
    val icon: String? = null,
    val trackId: Long = 0,
    val latLong: GeoPoint,
    val photoUrl: String? = null,
)

object WaypointReader {
    const val ID = "_id"
    const val NAME = "name" // waypoint name
    const val DESCRIPTION = "description" // waypoint description
    const val CATEGORY = "category" // waypoint category
    const val ICON = "icon" // waypoint icon
    const val TRACKID = "trackid" // track id
    const val LONGITUDE = "longitude" // longitude
    const val LATITUDE = "latitude" // latitude
    const val PHOTOURL = "photoUrl" // url for the photo

    val PROJECTION = arrayOf(
        ID,
        NAME,
        DESCRIPTION,
        CATEGORY,
        ICON,
        TRACKID,
        LATITUDE,
        LONGITUDE,
        PHOTOURL
    )
    val NAME_PATTERN: Pattern = Pattern.compile("[+\\s]*\\((.*)\\)[+\\s]*$")
    val POSITION_PATTERN: Pattern = Pattern.compile(
        "([+-]?\\d+(?:\\.\\d+)?),\\s?([+-]?\\d+(?:\\.\\d+)?)"
    )
    val QUERY_POSITION_PATTERN: Pattern =
        Pattern.compile("q=([+-]?\\d+(?:\\.\\d+)?),\\s?([+-]?\\d+(?:\\.\\d+)?)")

    fun fromGeoUri(uri: String): Waypoint? {
        var schemeSpecific = uri.substring(uri.indexOf(":") + 1)

        var name: String? = null
        val nameMatcher = NAME_PATTERN.matcher(schemeSpecific)
        if (nameMatcher.find()) {
            name = URLDecoder.decode(nameMatcher.group(1), "UTF-8")
            if (name != null) {
                schemeSpecific = schemeSpecific.substring(0, nameMatcher.start())
            }
        }

        var positionPart = schemeSpecific
        var queryPart = ""
        val queryStartIndex = schemeSpecific.indexOf('?')
        if (queryStartIndex != -1) {
            positionPart = schemeSpecific.substring(0, queryStartIndex)
            queryPart = schemeSpecific.substring(queryStartIndex + 1)
        }

        val positionMatcher = POSITION_PATTERN.matcher(positionPart)
        var lat = 0.0
        var lon = 0.0
        if (positionMatcher.find()) {
            lat = positionMatcher.group(1)!!.toDouble()
            lon = positionMatcher.group(2)!!.toDouble()
        }

        val queryPositionMatcher = QUERY_POSITION_PATTERN.matcher(queryPart)
        if (queryPositionMatcher.find()) {
            lat = queryPositionMatcher.group(1)!!.toDouble()
            lon = queryPositionMatcher.group(2)!!.toDouble()
        }

        return Waypoint(latLong = GeoPoint(lat, lon), name = name)
    }

    /**
     * Reads the Waypoints from the Content Uri.
     */
    fun readWaypoints(
        resolver: ContentResolver,
        data: Uri,
        lastWaypointId: Long
    ): List<Waypoint> {
        return buildList {
            resolver.query(data, PROJECTION, null, null, null).use { cursor ->
                while (cursor!!.moveToNext()) {
                    readWaypointFromCursor(cursor, lastWaypointId)?.let(::add)
                }
            }
        }
    }

    private fun readWaypointFromCursor(cursor: Cursor, lastWaypointId: Long): Waypoint? {
        val waypointId = cursor.getLong(cursor.getColumnIndexOrThrow(ID))
        if (lastWaypointId > 0 && lastWaypointId >= waypointId) { // skip waypoints we already have
            return null
        }
        val name = cursor.getString(cursor.getColumnIndexOrThrow(NAME))
        val description =
            cursor.getString(cursor.getColumnIndexOrThrow(DESCRIPTION))
        val category = cursor.getString(cursor.getColumnIndexOrThrow(CATEGORY))
        val icon = cursor.getString(cursor.getColumnIndexOrThrow(ICON))
        val trackId = cursor.getLong(cursor.getColumnIndexOrThrow(TRACKID))
        val latitude =
            cursor.getInt(cursor.getColumnIndexOrThrow(LATITUDE)) / APIConstants.LAT_LON_FACTOR
        val longitude =
            cursor.getInt(cursor.getColumnIndexOrThrow(LONGITUDE)) / APIConstants.LAT_LON_FACTOR
        val photoUrl = cursor.getString(cursor.getColumnIndexOrThrow(PHOTOURL))

        return if (MapUtils.isValid(latitude, longitude)) {
            val latLong = GeoPoint(latitude, longitude)
            Waypoint(
                waypointId,
                name,
                description,
                category,
                icon,
                trackId,
                latLong,
                photoUrl
            )
        } else {
            null
        }
    }

    fun fromGeoIntent(intent: Intent, mapData: MapData) {
        require(intent.isGeoIntent())

        fromGeoUri(intent.data.toString())
            ?.let { waypoint ->
                mapData.addWaypointMarker(waypoint)
                mapData.updateMapPositionAndZoomLevel(waypoint.latLong, 15)
            }
    }
}

fun Intent.isGeoIntent() = "geo" == scheme && data != null
